crates/nu-path/src/lib.rs

```diff
@@ -0,0 +1,532 @@
+    let ret = match c {
+        Cow::Borrowed(path) => f(path.to_string_lossy()),
+        Cow::Owned(buf) => f(Cow::Owned(buf.to_string_lossy().to_string())),
+    };
+
+    match ret {
+        Cow::Borrowed(expanded) => Cow::Borrowed(Path::new(expanded)),
+        Cow::Owned(expanded) => Cow::Owned(PathBuf::from(expanded)),
+    }
+}
+
+const EXPAND_STR: &str = if cfg!(windows) { r"..\" } else { "../" };
+fn handle_dots_push(string: &mut String, count: u8) {
+    if count < 1 {
+        return;
+    }
+
+    if count == 1 {
+        string.push('.');
+        return;
+    }
+
+    for _ in 0..(count - 1) {
+        string.push_str(EXPAND_STR);
+    }
+
+    string.pop(); // remove last '/'
+}
+
+// Expands any occurence of more than two dots into a sequence of ../ (or ..\ on windows), e.g.
+// ... into ../..
+// .... into ../../../
+fn expand_ndots_string<'a>(path: Cow<'a, str>) -> Cow<'a, str> {
+    use std::path::is_separator;
+    // find if we need to expand any >2 dot paths and early exit if not
+    let mut dots_count = 0u8;
+    let ndots_present = {
+        for chr in path.chars() {
+            if chr == '.' {
+                dots_count += 1;
+            } else {
+                if is_separator(chr) && (dots_count > 2) {
+                    // this path component had >2 dots
+                    break;
+                }
+
+                dots_count = 0;
+            }
+        }
+
+        dots_count > 2
+    };
+
+    if !ndots_present {
+        return path;
+    }
+
+    let mut dots_count = 0u8;
+    let mut expanded = String::new();
+    for chr in path.chars() {
+        if chr == '.' {
+            dots_count += 1;
+        } else {
+            if is_separator(chr) {
+                // check for dots expansion only at path component boundaries
+                handle_dots_push(&mut expanded, dots_count);
+                dots_count = 0;
+            } else {
+                // got non-dot within path component => do not expand any dots
+                while dots_count > 0 {
+                    expanded.push('.');
+                    dots_count -= 1;
+                }
+            }
+            expanded.push(chr);
+        }
+    }
+
+    handle_dots_push(&mut expanded, dots_count);
+
+    expanded.into()
+}
+
+// Expands any occurence of more than two dots into a sequence of ../ (or ..\ on windows), e.g.
+// ... into ../..
+// .... into ../../../
+fn expand_ndots<'a>(path: Cow<'a, Path>) -> Cow<'a, Path> {
+    cow_map_path_str(path, expand_ndots_string)
+}
+
+pub fn absolutize<P, Q>(relative_to: P, path: Q) -> PathBuf
+where
+    P: AsRef<Path>,
+    Q: AsRef<Path>,
+{
+    let path = if path.as_ref() == Path::new(".") {
+        // Joining a Path with '.' appends a '.' at the end, making the prompt
+        // more ugly - so we don't do anything, which should result in an equal
+        // path on all supported systems.
+        relative_to.as_ref().to_owned()
+    } else if path.as_ref().starts_with("~") {
```

Comment:
I left the code here roughly as is but I don't understand why this is expands tilde on the `path`, shouldn't it be `relative_to` or am I misunderstanding?

Reply:
If I understand correctly (I didn't write the functions), the main path you want to absolutize is `path`. If the `path` is relative, it gets joined to `relative_to`. If `path` is absolute, it replaces `relative_to`, according to how `Path.join()` or [`PathBuf.push()` work in the standard library](https://doc.rust-lang.org/std/path/struct.PathBuf.html#method.push): "If `path` is absolute, it replaces the current path.".

Reply:
Aah ok, yeah that makes sense.